packages/keystone/adapterCache.js

```diff
@@ -0,0 +1,335 @@
+            logger.info(`ADAPTER_CACHE: Cache is NOT enabled for list: ${listName} -> No ${UPDATED_AT_FIELD} field`)
+        }
+
+        if (excludedLists.includes(listName) ||
+            !includeAllLists && !includedLists.includes(listName)) {
+            logger.info(`ADAPTER_CACHE: Cache is NOT enabled for list: ${listName} -> Cache is not included by config`)
+            continue
+        }
+
+        logger.info(`ADAPTER_CACHE: Cache is enabled for list: ${listName}`)
+
+        const originalItemsQuery = listAdapter._itemsQuery
+        listAdapter._itemsQuery = async ( args, opts ) => {
+
+            middleware.totalRequests++
+
+            let key = null
+
+            const argsJson = JSON.stringify(args)
+            if (argsJson !== '{}') {
+                key = listName + '_' + argsJson + '_' + stringifyComplexObj(opts)
+            }
+
+            let response = []
+            const cached = key ? cache[listName][key] : false
+            const tableLastUpdate = await middleware.getState(listName)
+
+            if (cached) {
+                const cacheLastUpdate = cached.lastUpdate
+                if (cacheLastUpdate && cacheLastUpdate === tableLastUpdate) {
+                    middleware.cacheHits++
+                    const cacheEvent = middleware.getCacheEvent({
+                        type: 'HIT',
+                        table: listName,
+                        key,
+                        result: JSON.stringify(cached.response),
+                    })
+                    middleware.writeChangeToHistory({ cache, event: cacheEvent, table: listName } )
+                    middleware.logEvent({ event: cacheEvent })
+                    return cloneDeep(cached.response)
+                }
+            }
+
+            response = await originalItemsQuery.apply(listAdapter, [args, opts] )
+
+            let copiedResponse = cloneDeep(response)
+
+            cache[listName][key] = {
+                lastUpdate: tableLastUpdate,
+                response: copiedResponse,
+            }
+
+            const cacheEvent = middleware.getCacheEvent({
+                type: 'MISS',
+                key,
+                table: listName,
+                result: JSON.stringify(copiedResponse),
+            })
+            middleware.writeChangeToHistory({ cache, event: cacheEvent, table: listName } )
+            middleware.logEvent({ event: cacheEvent })
+
+            return response
+        }
+
+        const originalUpdate = listAdapter.update
+        listAdapter.update = patchAdapterFunction(listName, 'UPDATE', originalUpdate, listAdapter, middleware)
+
+        const originalCreate = listAdapter.create
+        listAdapter.create = patchAdapterFunction(listName, 'CREATE', originalCreate, listAdapter, middleware)
+
+        const originalDelete = listAdapter.delete
+        listAdapter.delete = patchAdapterFunction(listName, 'DELETE', originalDelete, listAdapter, middleware)
```

Comment:
Is it necessary to cache `delete` queries?

Reply:
They are not cached, they are dropping lists if DELETE instruction is found